backend/utilities/pipelineQueryUtils.ts

```diff
@@ -0,0 +1,69 @@
+import { FilterQuery, Types } from "mongoose";
+import { GradingStatus } from "../models/testSession.model";
+
+type GroupResultsByIdReturnType = {
+  $group: {
+    _id: string;
+    averageScore: { $avg: string };
+    resultBreakdowns: { $push: string };
+  };
+};
+
+// Filter out tests that have the requested testId
+export const filterTestsByTestId = (
+  testId: string,
+): FilterQuery<Record<string, unknown>> => {
+  return {
+    $match: { test: { $eq: Types.ObjectId(testId) } },
+  };
+};
+
+// Filter out results that are not graded
+export const filterUngradedTests = {
+  $project: {
+    results: {
+      $filter: {
+        input: "$results",
+        as: "results",
+        cond: {
+          $eq: ["$$results.gradingStatus", GradingStatus.GRADED.toString()],
+        },
+      },
+    },
+    school: 1,
```

Comment:
hmm this school field is not actually needed for your `getSubmissionCountByTest` - it's also unrelated to filtering the ungraded tests. can you make this constant just have the filter? and then have the rest in the main file. i.e. for your function would be
```
      {
        $project: {
          results: filterUngradedTests,
        },
      },
```
and the others include the school field